akka-docs/rst/java/code/docs/http/javadsl/server/directives/BasicDirectivesExamplesTest.java

```diff
@@ -0,0 +1,788 @@
+    private Route apiRoute(Supplier<Route> innerRoutes) {
+      return mapResponse(this::nonSuccessToEmptyJsonEntity, innerRoutes);
+    }
+  }
+
+  @Test
+  public void testMapResponseAdvanced() {
+    //#mapResponse-advanced
+    final ApiRoute api = new ApiRoute(system());
+
+    final Route route = api.apiRoute(() ->
+      get(() -> complete(StatusCodes.INTERNAL_SERVER_ERROR))
+    );
+
+    // tests:
+    testRoute(route).run(HttpRequest.GET("/"))
+      .assertEntity("{}");
+    //#mapResponse-advanced
+  }
+
+  @Test
+  public void testMapRouteResult() {
+    //#mapRouteResult
+    // this directive is a joke, don't do that :-)
+    final Route route = mapRouteResult(r -> {
+      if (r instanceof akka.http.scaladsl.server.RouteResult.Complete) {
+        final HttpResponse response = ((akka.http.scaladsl.server.RouteResult.Complete) r).getResponse();
+        return new akka.http.scaladsl.server.RouteResult.Complete(
+          (akka.http.scaladsl.model.HttpResponse) response.withStatus(200)
+        );
+      } else {
+        return r;
+      }
+    }, () -> complete(StatusCodes.ACCEPTED));
+
+    // tests:
+    testRoute(route).run(HttpRequest.GET("/"))
+      .assertStatusCode(StatusCodes.OK);
+    //#mapRouteResult
+  }
+
+  @Test
+  public void testMapRouteResultFuture() {
+    //#mapRouteResultFuture
+    final Route route = mapRouteResultFuture(cr ->
+      cr.exceptionally(t -> {
+        if (t instanceof IllegalArgumentException) {
+          return new akka.http.scaladsl.server.RouteResult.Complete(
+            (akka.http.scaladsl.model.HttpResponse)
+              HttpResponse.create().withStatus(StatusCodes.INTERNAL_SERVER_ERROR));
+        } else {
+          return null;
+        }
+      }).thenApply(rr -> {
+        if (rr instanceof akka.http.scaladsl.server.RouteResult.Complete) {
+          final HttpResponse res = ((akka.http.scaladsl.server.RouteResult.Complete) rr).getResponse();
+          return new akka.http.scaladsl.server.RouteResult.Complete(
+            (akka.http.scaladsl.model.HttpResponse)
+              res.addHeader(Server.create(ProductVersion.create("MyServer", "1.0"))));
+        } else {
+          return rr;
+        }
+      }), () -> complete("Hello world!"));
+
+    // tests:
+    testRoute(route).run(HttpRequest.GET("/"))
+      .assertStatusCode(StatusCodes.OK)
+      .assertHeaderExists(Server.create(ProductVersion.create("MyServer", "1.0")));
+    //#mapRouteResultFuture
+  }
+
+//  @Test
+//  public void testMapRouteResult() {
+//    //#0mapRouteResult
+//    mapRouteResult(result ->)
+//    //#0mapRouteResult
+//  }
+
+  @Test
+  public void testMapResponseEntity() {
+    //#mapResponseEntity
+    final Route route = mapResponseEntity(entity -> {
+      if (entity instanceof HttpEntity.Strict) {
+        final HttpEntity.Strict strict = (HttpEntity.Strict) entity;
+        return HttpEntities.create(strict.getContentType(), ByteString.fromString("test").$plus$plus(strict.getData()));
```

Comment:
Use `ByteString.concat` instead of Scala API here. Linefeed to make it shorter would be nice.